File: Concurrency/1116_PrintZeroEvenOdd/ZeroEvenOdd.java
// Your runtime beats 100.00 % of java submissions

class ZeroEvenOdd
{
    private int n;
    private final Object monitor = new Object();
    private boolean firstDigit = true;
    private boolean isOdd = false;
    private boolean isEven = false;

    public ZeroEvenOdd(int n)
    {
        this.n = n;
    }

    // printNumber.accept(x) outputs "x", where x is an integer.
    public void zero(IntConsumer printNumber) throws InterruptedException
    {
        synchronized (monitor)
        {
            for (int i = 1; i <= n; ++i)
            {
                while (!firstDigit)
                {
                    monitor.wait();
                }

                // if (i < 10)
                // {
                printNumber.accept(0);
                // }

                if (i % 2 == 0)
                {
                    isEven = true;
                }
                else if (i % 2 == 1)
                {
                    isOdd = true;
                }

                firstDigit = false;
                monitor.notifyAll();
            }
        }
    }

    public void even(IntConsumer printNumber) throws InterruptedException
    {
        synchronized (monitor)
        {
            for (int i = 2; i <= n; i += 2)
            {
                while (!isEven)
                {
                    monitor.wait();
                }

                printNumber.accept(i);
                isEven = false;
                firstDigit = true;
                monitor.notifyAll();
            }
        }
    }

    public void odd(IntConsumer printNumber) throws InterruptedException
    {
        synchronized (monitor)
        {
            for (int i = 1; i <= n; i += 2)
            {
                while (!isOdd)
                {
                    monitor.wait();
                }

                printNumber.accept(i);
                isOdd = false;
                firstDigit = true;
                monitor.notifyAll();
            }
        }
    }
}

File: Concurrency/1116_PrintZeroEvenOdd/ZeroEvenOdd.cpp
// Runtime: 64 ms, faster than 27.87% of C++ online submissions for Print Zero Even Odd.
// Memory Usage: 9.3 MB, less than 100.00% of C++ online submissions for Print Zero Even Odd.

class ZeroEvenOdd
{
private:
    int n;
    std::mutex m;
    std::condition_variable cond;
    bool printZero = true;
    bool printEven = false;
    bool printOdd = false;

public:
    ZeroEvenOdd(int n)
    {
        this->n = n;
    }

    // printNumber(x) outputs "x", where x is an integer.
    void zero(function<void(int)> printNumber)
    {
        std::unique_lock<std::mutex> guard(m);
        for (int i = 1; i <= n; ++i)
        {
            cond.wait(guard, [&]{return printZero;});
            printNumber(0);
            printZero = false;
            if (i % 2 == 0)
            {
                printEven = true;
            }
            else
            {
                printOdd = true;
            }
            cond.notify_all();
        }
    }

    void even(function<void(int)> printNumber)
    {
        std::unique_lock<std::mutex> guard(m);
        for (int i = 2; i <= n; i += 2)
        {
            cond.wait(guard, [&]{return printEven;});
            printNumber(i);
            printEven = false;
            printZero = true;
            cond.notify_all();
        }
    }

    void odd(function<void(int)> printNumber)
    {
        std::unique_lock<std::mutex> guard(m);
        for (int i = 1; i <= n; i += 2)
        {
            cond.wait(guard, [&]{return printOdd;});
            printNumber(i);
            printOdd = false;
            printZero = true;
            cond.notify_all();
        }
    }
};
